Typst property output

gordonwoodhull opened the pull request:
Here is my work in progress on Typst property output. Supercedes #9623.

https://github.com/gordonwoodhull/pandoc/tree/typst-property-output

It handles 8 cases of attribute-bearing AST nodes:
* span text (no element properties because `text()` is the closest element to span)
* div and div text
* table and table text
* td and td text (3 cases)

Now that I have tests I would be comfortable implementing the rest, but I haven't needed them yet. Thought I'd check in to ask if I'm on track here. All suggestions are welcome!

I think the code is more idiomatic Haskell (but not perfect) and the usage is terse (except for the align case, see below). Some details of whitespace/line breaks probably need more work.

There aren't row elements in Typst; tr attributes would have to be underlaid on cell properties.

#### align is messy

There is one case that is very bad - the Typst reader extracts (horizontal) `align`, but this needs to be coalesced with `vertical-align` for the Typst cell `align`. An alternative design would be to not remove the style in the reader, and allow Lua to override the whole property. Technically I should also parse `typst:align` as addition and then dedupe, because Typst will error if there are duplicate terms in the align expression, but Lua only sees `vertical-align` currently, so it works out.

jgm commented on src/Text/Pandoc/Writers/Typst.hs:
```diff
@@ -88,6 +88,52 @@ pandocToTypst options (Pandoc meta blocks) = do
+
+formatTypstProps :: [(Text, Text)] -> [Text]
+formatTypstProps =
+  map (\(k,v) -> last (T.splitOn (T.pack ":") k) <> ": " <> v)
```
Avoid `last`, as it's a partial function.
Also you don't need `T.pack ":"`; `":"` will work by itself because we have OverloadedStrings.

gordonwoodhull replied:
Good points. I had to improve validation to drop keys like `typst:foo:bar`, so I took the opportunity to pull the last element at input time and not worry about invalid input during output.

jgm commented on test/Tests/Writers/Typst.hs:
```diff
@@ -0,0 +1,178 @@
+      , "  align(center)[#table("
+      , "    columns: 2,"
+      , "    align: (auto,auto,),"
+      , "    [#set text(fill: orange); Foo"
+      , ""
+      , "    ], [Bar"
+      , ""
+      , "    ],"
+      , "  )]"
+      , "  , kind: table"
+      , "  )"
+      ]
+    , "table cell text w attr" =:
+      let headers = []
+          rows = [Row nullAttr [
+              cellWith ("", [], [("typst:text:fill", "orange")]) AlignCenter (RowSpan 1) (ColSpan 1) $ para "Foo"
+            , simpleCell $ para "Bar"
+            ]]
+      in table
+                    emptyCaption
+                    [(AlignDefault,ColWidthDefault),(AlignDefault,ColWidthDefault)]
+                    (TableHead nullAttr $ toHeaderRow headers)
+                    [TableBody nullAttr 0 [] rows]
+                    (TableFoot nullAttr [])
+    =?> unlines [
+        "#figure("
+      , "  align(center)[#table("
+      , "    columns: 2,"
+      , "    align: (auto,auto,),"
+      , "    table.cell(align: center)[#set text(fill: orange); Foo"
+      , ""
+      , "    ], [Bar"
+      , ""
+      , "    ],"
+      , "  )]"
+      , "  , kind: table"
+      , "  )"
+      ]
+    ]
+  ]
```
I'd suggest just using a "command test" (see `test/commands/*.md`).
They are easier to create and much easier to maintain (since we can just do `make test TESTARGS=--accept` to update the expected output instead of doing it manually.

gordonwoodhull replied:
Aha, much simpler to write tests that way!

I threw in negative cases: elements that don't have `typst:*` attributes should not affect the output. This was covered by previous tests but worth it to be somewhat complete.

jgm commented on src/Text/Pandoc/Writers/Typst.hs:
```diff
@@ -88,6 +88,52 @@ pandocToTypst options (Pandoc meta blocks) = do
+toTypstSetTextUnchecked typstTextAttrs =
+  "#set text" <> parens (toTypstProps typstTextAttrs) <> "; " -- newline?
+
+-- there is probably a more idiomatic way to do this
+ifNotEmpty ::  ([a2] -> Doc Text) -> [a2] -> Doc Text
+ifNotEmpty f list =
+  case list of
+    [] -> ""
+    _ -> f list
+
```
I think it would be more elegant to use a fold.  E.g. `foldl' addTypstProperty mempty props`.
If the list `props` is empty, you get `mempty`, so you wouldn't need this ugly function.
It should be possible to make this code simpler.

gordonwoodhull replied:
It's a little bit more complicated, because we have this conditional text we want to add around the lists, like we want to emit either `#set text(prop:value); ` or the empty string at the top of a block where we are applying `typst:text:*` attributes. We don't want to emit a useless `#set text(); ` if that's even valid.

And likewise everywhere else: if a block has no attributes, we want to emit `#block[content]` and not `#block()[content]`

@cscheid and I looked at it together and agreed that most clear thing to do is not to try to abstract it, but just write functions that both check for the empty list and generate text. It's about half as long now.

gordonwoodhull commented:
Okay, I think I have addressed your concerns, as well as adding solid validation of `typst:*` keys.

Glad to make any other changes! Should I add documentation somewhere?

I think I'd prefer to wait to implement output of properties from other elements until we get more experience, unless you have any requests.

`<span>`, `<div>`, `<table>` and `<td>` are the most important elements as far as Quarto is concerned.

I can do another PR later with other elements as our users inevitably test the boundaries of this, or we find more uses for Typst properties.

Thanks for all your helpful comments so far.

jgm commented:
As for documentation, one possibility would be to have a separate document along the lines of
https://pandoc.org/org.html
whose source is `doc/org.md`.  Alternatively it could be a section in the manual like the current section for EPUBs, but my thinking is that this concerns a pretty niche feature that is only accessible through filters, so it probably is best to document it separately...

jgm commented on src/Text/Pandoc/Writers/Typst.hs:
```diff
@@ -88,6 +90,42 @@ pandocToTypst options (Pandoc meta blocks) = do
+pickTypstAttrs =
+  partitionEithers . catMaybes . map (\(k,v) ->
+    let spl = T.splitOn ":" k in
+    if length(spl) < 2 || length(spl) > 3
+      || head spl /= "typst"
+      || (length(spl) == 3 && spl !! 1 /= "text")
```
We like to avoid using partial functions like `head`. I think it's cleaner to use pattern matching for this kind of thing. Could you do something like
```hs

case T.splitOn ":" k of
  "typst":"text":x -> Right (x,v)
  "typst":x -> Left (x, v)
  _ -> Nothing
```

Second thought: instead of using `partitionEithers . catMaybes . map`, why not just use a fold that directly constructs the tuple?  This way you also avoid needing to use Maybe or Either.

```hs
pickTypstAttrs = foldr go ([],[])
  where
    go (k,v) (eltProps, textProps) =
      case T.splitOn ":" k of
        "typst":"text":x -> (eltProps, (x,v):textProps)
        "typst":x -> ((x,v):eltProps, textProps)
        _ -> (eltProps, textProps)
```
Or even more compactly with Data.Bifunctor.(first,second)
```hs
pickTypstAttrs = foldr go ([],[])
  where
    go (k,v)
      case T.splitOn ":" k of
        "typst":"text":x -> second ((x,v):)
        "typst":x -> first ((x,v):)
        _ -> id
```

gordonwoodhull replied:
This is very nice!

All five of my logical conditions covered by three patterns, and the meaning is very clear with the bifunctor approach.

jgm commented on src/Text/Pandoc/Writers/Typst.hs:
```diff
@@ -88,6 +90,42 @@ pandocToTypst options (Pandoc meta blocks) = do
+       then Just $ Left (spl !! 1, v)
+       else Just $ Right (spl !! 2, v))
+
+formatTypstProps :: [(Text, Text)] -> [Text]
+formatTypstProps =
+  map (\(k,v) -> k <> ": " <> v)
+
+toTypstPropsListSep :: [(Text, Text)] -> Doc Text
+toTypstPropsListSep typstAttrs =
+   literal (T.intercalate ", " $ formatTypstProps typstAttrs)
+
```
With this approach the attribute lists will be unbreakable.
That's probably not what we want.  Better:
```hs
formatTypstProps = map (\(k,v) -> literal (k <> ": " <> v))
toTypstPropsListSep = hsep . intersperse "," . formatTypstProps
```

gordonwoodhull replied:
Great, knew I was missing something about line breaks. Thanks!

jgm commented on src/Text/Pandoc/Writers/Typst.hs:
```diff
@@ -88,6 +90,42 @@ pandocToTypst options (Pandoc meta blocks) = do
+
+toTypstTextElement :: [(Text, Text)] -> Doc Text -> Doc Text
+toTypstTextElement [] content = content
+toTypstTextElement typstTextAttrs content = "#text" <> parens (toTypstPropsListSep typstTextAttrs) <> brackets content
```
Could simplify by using `toTypstPropsListParens` here instead of the spelled out version

gordonwoodhull commented:
Excellent feedback - I think it is very clean and simple now.

Working on docs.

iandol commented:
Amazingly useful contribution @gordonwoodhull!!! 🥳

jgm commented on doc/typst-property-output.md:
````diff
@@ -0,0 +1,155 @@
+the command
+
+```sh
+quarto pandoc -f html -t typst --lua-filter ./typst-property-example.lua
````
Since this is pandoc's documentation, not quarto, shouldn't this just be `pandoc`?

gordonwoodhull replied:
Yep!

gordonwoodhull commented:
I fixed the example, did a few more copy edits, and squashed everything into one commit. 

I think this is ready! Please let me know if you want any further changes.

jgm commented on test/command/typst-property-output.md:
````diff
@@ -0,0 +1,153 @@
+```
+
+```
+% pandoc -f html -t typst
````
Is there a difference between this test and the one at line 76?

jgm replied:
Did you mean to have a test of an attribute on `table`?

gordonwoodhull replied:
Hmm, I guess I added negative tests for both table and cell without realizing they are the same. Removed.

gordonwoodhull commented:
I found that I had failed to update the ugly special case for `align` when I improved validation, so I pushed a fix and associated test.

Also I noticed that we can use a set-text rule instead of text element for table, since the table is inside the figure & align scopes and the set-text rule won't leak out.

jgm commented:
I think it's good to go! Thanks.

gordonwoodhull commented:
Thank you!

gordonwoodhull commented:
This is working great and we'd like to include "Typst CSS" in Quarto 1.5.

When you're around and have a minute, please release 3.1.14!
